pandas/io/stata.py

```diff
@@ -951,222 +953,324 @@ def __init__(self, path_or_buf, convert_dates=True,
-                                for i in range(self.nvar)]
-            else:
-                self.vlblist = [self._null_terminate(self.path_or_buf.read(32))
-                                for i in range(self.nvar)]
 
-            # ignore expansion fields (Format 105 and later)
-            # When reading, read five bytes; the last four bytes now tell you
-            # the size of the next read, which you discard.  You then continue
-            # like this until you read 5 bytes of zeros.
+    def _read_117_header(self):
+        self.nobs = struct.unpack(self.byteorder + 'I',
+                                  self.path_or_buf.read(4))[0]
+        self.path_or_buf.read(11)  # </N><label>
+        strlen = struct.unpack('b', self.path_or_buf.read(1))[0]
+        self.data_label = self._null_terminate(self.path_or_buf.read(strlen))
+        self.path_or_buf.read(19)  # </label><timestamp>
+        strlen = struct.unpack('b', self.path_or_buf.read(1))[0]
+        self.time_stamp = self._null_terminate(self.path_or_buf.read(strlen))
+        self.path_or_buf.read(26)  # </timestamp></header><map>
+        self.path_or_buf.read(8)  # 0x0000000000000000
+        self.path_or_buf.read(8)  # position of <map>
+        seek_vartypes = struct.unpack(
+            self.byteorder + 'q', self.path_or_buf.read(8))[0] + 16
+        seek_varnames = struct.unpack(
+            self.byteorder + 'q', self.path_or_buf.read(8))[0] + 10
+        seek_sortlist = struct.unpack(
+            self.byteorder + 'q', self.path_or_buf.read(8))[0] + 10
+        seek_formats = struct.unpack(
+            self.byteorder + 'q', self.path_or_buf.read(8))[0] + 9
+        seek_value_label_names = struct.unpack(
+            self.byteorder + 'q', self.path_or_buf.read(8))[0] + 19
+
+        self.path_or_buf.read(8)  # <variable_lables>, throw away
+        # Stata 117 data files do not follow the described format.  This is
+        # a work around that uses the previous label, 33 bytes for each
+        # variable, 20 for the closing tag and 17 for the opening tag
+        seek_variable_labels = seek_value_label_names + (33*self.nvar) + 20 + 17
+        self.path_or_buf.read(8)  # <characteristics>
+        self.data_location = struct.unpack(
+            self.byteorder + 'q', self.path_or_buf.read(8))[0] + 6
+        self.seek_strls = struct.unpack(
+            self.byteorder + 'q', self.path_or_buf.read(8))[0] + 7
+        self.seek_value_labels = struct.unpack(
+            self.byteorder + 'q', self.path_or_buf.read(8))[0] + 14
+        #self.path_or_buf.read(8)  # </stata_dta>
+        #self.path_or_buf.read(8)  # EOF
+        self.path_or_buf.seek(seek_vartypes)
+        typlist = [struct.unpack(self.byteorder + 'H',
+                                 self.path_or_buf.read(2))[0]
+                   for i in range(self.nvar)]
+        self.typlist = [None]*self.nvar
+        try:
+            i = 0
+            for typ in typlist:
```

Comment:
further you its better to make the typlist conversion as a function itself

e.g. ``self.dtypelist = self._get_dtypes_from_typelist(dtyplist)`` (then can use it in 117 and 118) versions